cmake/test/src/cmake.py

```diff
@@ -7,158 +7,143 @@
+    make_targets=None,
+    install_directory=None,
 ):
-    """
-    Run the CMake command, and any number of make commands. Ensures that the calls all process as
-    expected, and that the expected outputs are produced.
-
-    :param build_path: target path. Usually <FPRIME>/Ref. <FPRIME> replaced with f prime dir.
-    :param expected_outputs: list of files to check for. Relative to build directory.
-    :param build_directory: directory in which to perform the build, default is a temporary dir.
-    :param make_targets: list of make targets to run.
-    :param options: options to supply to the cmake system.
-    :return: True if the test passes, False if there is an error
-    """
-    start_time = time.time()
-    if build_directory is not None:
-        raise Exception("Named build directories not supported, yet.")
-    owd = os.getcwd()
-    tmpd = tempfile.mkdtemp()
-
-    if not "FPRIME_INSTALL_DEST" in options:
-        options["FPRIME_INSTALL_DEST"] = os.path.join(tmpd, "build-artifacts")
-
-    try:
-        os.chdir(tmpd)
-        build_path = build_path.replace(
-            "<FPRIME>", os.path.join(os.path.dirname(__file__), "..", "..", "..")
-        )
-        assert run_cmake(
-            build_path, options=options
-        ), "CMake failed for path '{0}' and options '{1}'".format(build_path, options)
-        for target in make_targets:
-            assert run_make(
-                target
-            ), "Failed to build '{0}' target for '{1}' and options '{2}'".format(
-                target, build_path, options
-            )
-        assert_exists(expected_outputs, options["FPRIME_INSTALL_DEST"], start_time)
-    finally:
-        os.chdir(owd)
-        shutil.rmtree(tmpd, ignore_errors=False)
+    """Generate and build a cmake deployment, then returns a pytest fixture for it"""
+    cmake_arguments = {} if cmake_arguments is None else cmake_arguments
+    build_directory = Path(tempfile.mkdtemp())
+    install_directory_calc = Path(
+        Path(source_directory) / "build-artifacts"
+        if install_directory is None
+        else install_directory
+    )
+    make_targets = ["all"] if make_targets is None else make_targets
+    if install_directory is not None:
+        cmake_arguments["CMAKE_INSTALL_PREFIX"] = str(install_directory)
+
+    cmake_output = run_cmake(source_directory, build_directory, cmake_arguments)
+    target_outputs = {
+        target: run_make(build_directory, target) for target in make_targets
+    }
+
+    @pytest.fixture(scope="session", name=fixture_name)
+    def fixture_function():
+        yield {
+            "source": source_directory,
+            "build": build_directory,
+            "install": install_directory_calc,
+            "cmake": cmake_output,
+            "targets": target_outputs,
+        }
+        # shutil.rmtree(build_directory, ignore_errors=True)
+        # shutil.rmtree(install_directory, ignore_errors=True)
+
+    return fixture_function
```

Comment:
Is there a reason you're using a fixture instead of a function return value? In general I consider dyanamic dependency injection a bit of an antipattern. Dependency injunction can be complicated enough by itself, but dynamically creating dependencies can make tracing the injection process nearly impossible. Can you tell I've gotten burned by some crazy java dependency injection environments in the past? 😂 

This is a very simple use case of dependency injection so I'm not too concerned, but I do wonder if there's a better approach that could be used.

Are you returning the fixture to prevent it from going out of scope?

Reply:
I am using fixtures because `pytest` will handle the cleanup once all the tests are run.  I don't have to worry about adding `.atexit` calls and whatnot to ensure that the fixture has been cleaned-up.  In addition, this allows the reuse of test fixtures without forcing a separate rebuild.

The only caveat is that the fixture must exist in the module using it....hence the return call.  The fixture is returned such that it exists in the calling module.

Reply:
I might be overlooking it, but is there any explicit cmake cleanup occurring right now? Either way I'm not too concerned since the build is occurring in temporary directories. In some ways I actually prefer the lack of explicit cleanup so I can poke around the temporary build directory on test failure.

Reply:
It is those two lines that are commented out.  I need to reenable them.  Good catch!

Reply:
Ok, reenabled.